File: Task/Resistance-calculator/C++/resistance-calculator.cpp
#include <cstdint>
#include <iomanip>
#include <iostream>
#include <map>
#include <memory>
#include <numeric>
#include <sstream>
#include <stack>
#include <string>
#include <vector>

class Resistor {
public:
	Resistor(const char& symbol, const double& resistance,
			 const std::shared_ptr<Resistor>& a, const std::shared_ptr<Resistor>& b)
	: symbol(symbol), resistance(resistance), voltage(0.0), a(a), b(b) {}

	Resistor(const char& symbol, const double& resistance)
	: symbol(symbol), resistance(resistance), voltage(0.0) {}

	char symbol;
	double resistance;
	double voltage;
	std::shared_ptr<Resistor> a;
	std::shared_ptr<Resistor> b;
};

double get_resistance(const std::shared_ptr<Resistor>& resistor) {
	switch ( resistor->symbol ) {
		case '+' : return get_resistance(resistor->a) + get_resistance(resistor->b);
		case '*' : return 1.0 / ( 1.0 / get_resistance(resistor->a) + 1.0 / get_resistance(resistor->b) );
		default  : return resistor->resistance;
	}
}

void set_voltage(const std::shared_ptr<Resistor>& resistor, const double& voltage) {
    switch ( resistor->symbol ) {
    	case '+' : {
    		const double ra = get_resistance(resistor->a);
    		const double rb = get_resistance(resistor->b);
    		set_voltage(resistor->a, ra / ( ra + rb ) * voltage);
        	set_voltage(resistor->b, rb / ( ra + rb ) * voltage);
        	resistor->voltage = resistor->a->voltage + resistor->b->voltage;
    		break;
    	}
    	case '*' : {
    		set_voltage(resistor->a, voltage);
        	set_voltage(resistor->b, voltage);
        	resistor->voltage = voltage;
        	break;
    	}
    	default : resistor->voltage = voltage;
    }
}

double get_current(const std::shared_ptr<Resistor>& resistor) {
    return resistor->voltage / get_resistance(resistor);
}

double get_power(const std::shared_ptr<Resistor>& resistor) {
    return resistor->voltage * get_current(resistor);
}

void report(const std::shared_ptr<Resistor>& resistor, const std::string& level) {
    std::cout << std::fixed << std::setprecision(3) << std::setw(8) << get_resistance(resistor)
    		  << std::setw(8) << resistor->voltage << std::setw(8) << get_current(resistor)
			  << std::setw(8) << get_power(resistor) << level << resistor->symbol << std::endl;
    if ( resistor->a ) {
        report(resistor->a, level + "| ");
    }
    if ( resistor->b ) {
        report(resistor->b, level + "| ");
    }
}

void report(const std::shared_ptr<Resistor>& resistor) {
	report(resistor, " ");
}

bool is_digit(const char& ch) {
    return ch >= '0' && ch <= '9';
}

std::string convert_to_RPN(const std::string& text) {
	std::map<char, uint32_t> precedence = { { '(', 0 }, { '+', 1 }, { '*', 2 } };

	std::vector<std::string> items{ };
	std::stack<char> stack{ };
	uint32_t end = 0;
	while ( end < text.length() ) {
		uint32_t start = end;
		const char ch = text[end];
		end += 1;
		if ( ch == '(' ) {
			stack.push(ch);
		} else if ( ch == '+' || ch == '*' ) {
			while ( ! stack.empty() && precedence[ch] <= precedence[stack.top()] ) {
				items.emplace_back(std::string(1, stack.top()) + " "); stack.pop();
			}
			stack.push(ch);
		} else if ( ch == ')' ) {
			while ( stack.top() != '(' ) {
				items.emplace_back(std::string(1, stack.top()) + " "); stack.pop();
			}
			stack.pop();
		} else {
			if ( is_digit(ch) ) {
				while ( end < text.length() && is_digit(text[end]) ) {
					end += 1;
				}
				items.emplace_back(text.substr(start, end - start) + " "); // Extract a multi-digit number
			}
		}
	}

	while ( ! stack.empty() ) {
		items.emplace_back(std::string(1, stack.top()) + " "); stack.pop();
	}
	return std::accumulate(items.begin(), items.end(), std::string{ });
}

std::vector<std::string> split_string(const std::string& text, const char& delimiter) {
	std::vector<std::string> words;
	std::istringstream stream(text);
	std::string word;
	while ( std::getline(stream, word, delimiter) ) {
	    words.emplace_back(word);
	}
    return words;
}

std::shared_ptr<Resistor> build(const std::string& text) {
    std::stack<std::shared_ptr<Resistor>> stack{ };
    for ( const std::string& word : split_string(text, ' ') ) {
        if ( word == "+" ) {
			std::shared_ptr<Resistor> b = stack.top(); stack.pop();
			std::shared_ptr<Resistor> a = stack.top(); stack.pop();
			std::shared_ptr<Resistor> resistor = std::make_shared<Resistor>( '+', 0.0, a, b );
			stack.push(resistor);
		} else if ( word == "*" ) {
        	std::shared_ptr<Resistor> b = stack.top(); stack.pop();
			std::shared_ptr<Resistor> a = stack.top(); stack.pop();
			std::shared_ptr<Resistor> resistor = std::make_shared<Resistor>( '*', 0.0, a, b );
			stack.push(resistor);
        } else {
        	const double value = std::stod(word);
        	std::shared_ptr<Resistor> resistor = std::make_shared<Resistor>( 'r', value );
        	stack.push(resistor);
        }
    }

    std::shared_ptr<Resistor> result = stack.top(); stack.pop();
    return result;
}

void calculator(const std::string& title, const std::string& resistors, const double& voltage) {
	std::cout << title << std::endl;
	std::shared_ptr<Resistor> root = build(resistors);
	std::cout << "    Ohm    Volt   Ampere   Watt  Network tree" << std::endl;
	set_voltage(root, voltage);
	report(root);
}

int main() {
	calculator("RPN syntax:", "10 2 + 6 * 8 + 6 * 4 + 8 * 4 + 8 * 6 +", 18.0);
	std::cout << std::endl;

	calculator("Infix syntax:", convert_to_RPN("((((10 + 2) * 6 + 8) * 6 + 4) * 8 + 4) * 8 + 6"), 18.0);
}

File: Task/Resistance-calculator/Java/resistance-calculator.java
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.Stack;

public final class ResistanceCalculator {

	public static void main(String[] args) {
		rpnSyntax("RPN syntax:", "10 2 + 6 * 8 + 6 * 4 + 8 * 4 + 8 * 6 +", 18.0);
		System.out.println();
	
		infixSyntax("Infix syntax:", "((((10 + 2) * 6 + 8) * 6 + 4) * 8 + 4) * 8 + 6", 18.0);	
	}
	
	private static void rpnSyntax(String title, String rpn, double voltage) {
		System.out.println(title);
		Resistor root = build(rpn);
		System.out.println("    Ohm    Volt   Ampere   Watt  Network tree");
		root.setVoltage(voltage);
		root.report();
	}
	
	private static void infixSyntax(String title, String infix, double voltage) {
		rpnSyntax(title, convertToRPN(infix), voltage);
	}
	
	private static Resistor build(String text) {
		Stack<Resistor> stack = new Stack<Resistor>();
		for ( String word : text.split(" ") ) {
			if ( word.equals("+") ) {
				stack.push( new SerialResistor(stack.pop(), stack.pop()) );
			} else if ( word.equals("*") ) {
				stack.push( new ParallelResistor(stack.pop(), stack.pop()) );
			} else {
				stack.push( new Resistor(Double.valueOf(word)) );
			}
		}
		return stack.pop();
	}
	
	private static String convertToRPN(String text) {
	    final Map<Character, Integer> precedence = Map.ofEntries(
	    	Map.entry('(', 0), Map.entry('+', 1), Map.entry('*', 2) );

	    List<String> items = new ArrayList<String>();
	    Stack<Character> stack = new Stack<Character>();
	    int end = 0;
	    while ( end < text.length() ) {
	        int start = end;
	        final char ch = text.charAt(end);
	        end += 1;
	        switch ( ch ) {
	        	case '(' -> stack.push(ch);
		        case '+', '*' -> {
		        	while ( ! stack.isEmpty() && precedence.get(ch) <= precedence.get(stack.peek()) ) {
		                items.add(String.valueOf(stack.pop()));
		        	}
		            stack.push(ch);
		        }
		        case ')' -> {
		        	while ( stack.peek() != '(' ) {
		                items.add(String.valueOf(stack.pop()));
		            }
		            stack.pop();
		        }
		        default -> {
		        	if ( Character.isDigit(ch) ) {
			            while ( end < text.length() && Character.isDigit(text.charAt(end)) ) {
			                end += 1;
			            }
			            items.add(text.substring(start, end)); // Extract a multi-digit number
			        }
		        }
	        }
	    }
	
	    while ( ! stack.isEmpty() ) {
	    	items.add(String.valueOf(stack.pop()));
	    } 	
	    return String.join(" ", items);
	}

}

class Resistor {
	
	public Resistor(char aSymbol, double aResistance, Resistor aA, Resistor aB) {
		symbol = aSymbol;
		resistance = aResistance;
		a = aA;
		b = aB;
	}
	
	public Resistor(double aResistance) {
		this('r', aResistance, null, null);
	}
	
	protected double getResistance() { return resistance; }
	protected double getVoltage() { return voltage; }
	protected void setVoltage(double aVoltage) { voltage = aVoltage; }
	protected double getCurrent() { return voltage / resistance; }
	protected double getPower() { return getCurrent() * voltage; }
	
	protected void report() { report(""); }
	protected void report(String level) {
		System.out.println(String.format("%8.3f%8.3f%8.3f%8.3f%s %c",
			getResistance(), getVoltage(), getCurrent(), getPower(), level, symbol));
		if ( a != null ) { a.report(level + " | "); }
		if ( b != null ) { b.report(level + " | "); }
	}	
	
	private char symbol;
	private double resistance;
	private double voltage;
	private Resistor a;
	private Resistor b;

}

final class SerialResistor extends Resistor {
	
	public SerialResistor(Resistor aA, Resistor aB) {		
		super('+', 0.0, aB, aA);
		a = aA;
		b = aB;
	}
	
	protected double getResistance() { return a.getResistance() + b.getResistance(); }
	protected double getVoltage() { return voltage; }
	protected void setVoltage(double aVoltage) {
		final double resistanceA = a.getResistance();
		final double resistanceB = b.getResistance();
		a.setVoltage(aVoltage * resistanceA / ( resistanceA + resistanceB ));
		b.setVoltage(aVoltage * resistanceB / ( resistanceA + resistanceB ));
		voltage = aVoltage;
	}
	protected double getCurrent() { return voltage / getResistance(); }
	protected double getPower() { return getCurrent() * voltage; }
	
	private double voltage;
	private Resistor a;
	private Resistor b;
	
}

final class ParallelResistor extends Resistor {
	
	public ParallelResistor(Resistor aA, Resistor aB) {		
		super('*', 0.0, aB, aA);
		a = aA;
		b = aB;
	}
	
	protected double getResistance() { return 1.0 / ( 1.0 / a.getResistance() + 1.0 / b.getResistance() ); }
	protected double getVoltage() { return voltage; }
	protected void setVoltage(double aVoltage) {
		a.setVoltage(aVoltage);
		b.setVoltage(aVoltage);
		voltage = aVoltage;
	}
	protected double getCurrent() { return voltage / getResistance(); }
	protected double getPower() { return getCurrent() * voltage; }
	
	private double voltage;
	private Resistor a;
	private Resistor b;
	
}
